akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -137,6 +144,129 @@ object GraphStageLogic {
+          }
+      }
+
+    @tailrec final def remWatcher(watchee: ActorRef, watcher: ActorRef): Unit = {
+      _watchedBy.get() match {
+        case StageTerminatedTombstone ⇒ // do nothing...
+        case watchedBy ⇒
+          val watcheeSelf = watchee == this
+          val watcherSelf = watcher == this
+
+          if (watcheeSelf && !watcherSelf) {
+            if (watchedBy.contains(watcher))
+              if (!_watchedBy.compareAndSet(watchedBy, watchedBy - watcher))
+                remWatcher(watchee, watcher) // try again
+          } else if (!watcheeSelf && watcherSelf) {
+            unwatch(watchee)
+          } else {
+            log.error("BUG: illegal Unwatch(%s,%s) for %s".format(watchee, watcher, this))
+          }
+      }
+    }
+
+    def watch(actorRef: ActorRef): Unit =
+      actorRef.asInstanceOf[InternalActorRef].sendSystemMessage(Watch(actorRef.asInstanceOf[InternalActorRef], this))
+
+    def unwatch(actorRef: ActorRef): Unit =
+      actorRef.asInstanceOf[InternalActorRef].sendSystemMessage(Unwatch(actorRef.asInstanceOf[InternalActorRef], this))
+  }
+
+  /**
+   * Minimal actor to work with other actors and watch them in a synchronous ways
+   */
+  class StageActorRef(val provider: ActorRefProvider, val log: LoggingAdapter,
+                      getAsyncCallback: StageActorRef.Receive ⇒ AsyncCallback[(ActorRef, Any)],
+                      initialReceive: StageActorRef.Receive,
+                      override val path: ActorPath) extends akka.actor.MinimalActorRef with MinimalDeathWatch {
+
+    val callback = getAsyncCallback(internalReceive)
+
+    @volatile
+    private[this] var behaviour = initialReceive
+
+    /** INTERNAL API */
+    private[akka] def internalReceive(pack: (ActorRef, Any)): Unit =
+      behaviour(pack)
+
+    override def !(message: Any)(implicit sender: ActorRef = Actor.noSender): Unit = {
+      message match {
+        case m @ (PoisonPill | Kill) ⇒
+          log.warning("{} message sent to StageActorRef({}) will be ignored, since it is not a real Actor." +
+            "Use a custom message type to communicate with it instead.", m, path)
+        case _ ⇒
+          callback.invoke((sender, message))
+      }
+    }
+
+    override def sendSystemMessage(message: SystemMessage): Unit = message match {
+      case w: Watch   ⇒ addWatcher(w.watchee, w.watcher)
+      case u: Unwatch ⇒ remWatcher(u.watchee, u.watcher)
+      case DeathWatchNotification(actorRef, _, _) ⇒
+        this.!(Terminated(actorRef)(existenceConfirmed = true, addressTerminated = false))
+      case _ ⇒ //ignore all other messages
+    }
+
+    /**
+     * Special `become` allowing to swap the behaviour of this StageActorRef.
+     * Unbecome is not available.
+     */
+    def become(receive: StageActorRef.Receive): Unit = {
+      behaviour = receive
+    }
+  }
+  object StageActorRef {
+    type Receive = ((ActorRef, Any)) ⇒ Unit
```

Comment:
No, you only need to register a prox function instead of the callback that just unwraps the pair. It helps people to directly call `match` on the message and not the pair.